jdaviz/core/template_mixin.py

```diff
@@ -230,32 +302,50 @@ def _update_has_subregions(self):
     def labels(self):
         return [s['label'] for s in self.items if 'label' in s.keys()]
 
+    @cached_property
+    def selected_item(self):
+        for item in self.items:
+            if item['label'] == self.selected:
+                return item
+        return {}
+
     @cached_property
     def selected_obj(self):
-        if self.selected == "Entire Spectrum":
+        if self.selected == self._default_text:
             return None
-        return self.app.get_subsets_from_viewer("spectrum-viewer",
-                                                subset_type="spectral").get(self.selected)
+        subset_type = self.selected_item['type']
+        # NOTE: we use reference names here instead of IDs since get_subsets_from_viewer requires
+        # that.  For imviz, this will mean we won't be able to loop through each of the viewers,
+        # but the original viewer should have access to all the subsets.
+        for viewer_ref in self.viewer_refs:
+            match = self.app.get_subsets_from_viewer(viewer_ref,
```

Comment:
Aperture photometry still uses your logic to actually extract the object (although in the future I think it would be nice to migrate to use the logic in the new component), and the dropdown is populated by messages, not looping over layers.  If you can find a reproducible case that shows an issue, let me know (although I'd argue that we should fix `get_data_from_viewer` if it is failing to show items outside viewer limits or something like that... but we definitely want to avoid re-introducing broken behavior in this PR).